build-tools/packages/build-tools/src/typeValidator/packageJson.ts

```diff
@@ -131,13 +145,155 @@ export async function getPackageDetails(packageDir: string): Promise<PackageDeta
+                prevVersion = getVersionRange(baseMinor, "~");
+            } else {
+                prevVersion = `~${sv.major}.${sv.minor}.0`;
+            }
+
+            break;
+        }
+
+        case "previousMajor": {
+            if (previousMajorVersion === undefined) {
+                throw new Error(`Previous major version is undefined.`);
+            }
+
+            prevVersion = previousMajorVersion;
+            break;
+        }
+
+        case "previousMinor": {
+            if (previousMinorVersion === undefined) {
+                throw new Error(`Previous minor version is undefined.`);
+            }
+
+            prevVersion = previousMinorVersion;
+            break;
+        }
+
+        case "previousPatch": {
+            if (previousPatchVersion === undefined) {
+                throw new Error(`Previous patch version is undefined.`);
+            }
+
+            prevVersion = previousPatchVersion;
+            break;
+        }
+
+        case "^previousMajor": {
+            if (previousMajorVersion === undefined) {
+                throw new Error(`Previous major version is undefined.`);
+            }
+
+            prevVersion = isInternalVersionScheme(previousMajorVersion)
+                ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                  getVersionRange(previousMajorVersion!, "^")
+                : `^${previousMajorVersion}`;
+            break;
+        }
+
+        case "^previousMinor": {
+            if (previousMinorVersion === undefined) {
+                throw new Error(`Previous minor version is undefined.`);
+            }
+
+            prevVersion = isInternalVersionScheme(previousMinorVersion)
+                ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                  getVersionRange(previousMinorVersion!, "^")
+                : `^${previousMinorVersion}`;
+            break;
+        }
+
+        case "~previousMajor": {
+            if (previousMajorVersion === undefined) {
+                throw new Error(`Previous major version is undefined.`);
+            }
+
+            prevVersion = isInternalVersionScheme(previousMajorVersion)
+                ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                  getVersionRange(previousMajorVersion!, "~")
+                : `~${previousMajorVersion}`;
+            break;
+        }
+
+        case "~previousMinor": {
+            if (previousMinorVersion === undefined) {
+                throw new Error(`Previous minor version is undefined.`);
+            }
+
+            prevVersion = isInternalVersionScheme(previousMinorVersion)
+                ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                  getVersionRange(previousMinorVersion!, "~")
+                : `~${previousMinorVersion}`;
+            break;
+        }
+    }
```

Comment:
We don't have a default case here. What should this return if the user input a malformed constraint string?